common/djangoapps/embargo/api.py

```diff
@@ -0,0 +1,266 @@
+
+    """
+    cache_key = u'user.{user_id}.profile.country'.format(user_id=user.id)
+    profile_country = cache.get(cache_key)
+    if profile_country is None:
+        profile = getattr(user, 'profile', None)
+        if profile is not None and profile.country.code is not None:
+            profile_country = profile.country.code.upper()
+        else:
+            profile_country = ""
+        cache.set(cache_key, profile_country)
+
+    # with whitelist if returns false return error
+    if not CountryAccessRule.is_course_embargoed_in_country_list(course_id, profile_country, WHITE_LIST):
+        return REASONS['profile_country'].format(
+            user_id=unique_id_for_user(user),
+            profile_country=profile_country,
+            from_course=_from_course_msg(course_id, course_is_embargoed)
+        )
+
+    # with blacklist if returns true return error
+    if CountryAccessRule.is_course_embargoed_in_country_list(course_id, profile_country, BLACK_LIST):
+        return REASONS['profile_country'].format(
+            user_id=unique_id_for_user(user),
+            profile_country=profile_country,
+            from_course=_from_course_msg(course_id, course_is_embargoed)
+        )
+
+
+def _country_code_from_ip(ip_addr):
+    """
+    Return the country code associated with an IP address.
+    Handles both IPv4 and IPv6 addresses.
+
+    Args:
+        ip_addr (str): The IP address to look up.
+
+    Returns:
+        str: A 2-letter country code.
+
+    """
+    if ip_addr.find(':') >= 0:
+        return pygeoip.GeoIP(settings.GEOIPV6_PATH).country_code_by_addr(ip_addr)
+    else:
+        return pygeoip.GeoIP(settings.GEOIP_PATH).country_code_by_addr(ip_addr)
+
+
+def check_access(user, ip_address, course_key):
+    """
+    Check is the user with this ip_address has access to the given course
+
+    Params:
+        user (User): Currently logged in user object
+        ip_address (str): The ip_address of user
+        course_key (CourseLocator): CourseLocator object the user is trying to access
+
+    Returns:
+        Redirect to a URL configured in Django settings or a forbidden response if any constraints fails or None
+
+    """
+    course_is_restricted = RestrictedCourse.is_restricted_course(course_key)
+    # If they're trying to access a course that cares about embargoes
+    if course_is_restricted:
+
+        # Construct the list of functions that check whether the user is embargoed.
+        # We wrap each of these functions in a decorator that logs the reason the user
+        # was blocked.
+        # Each function should return `True` iff the user is blocked by an embargo.
+        check_functions = [
+            _log_embargo_reason(check_func, course_key, course_is_restricted)
+            for check_func in [
+                partial(_is_embargoed_by_ip, ip_address, course_id=course_key),
+                partial(_is_embargoed_by_profile_country, user, course_id=course_key)
+            ]
+        ]
```

Comment:
I originally set up the "check function" structure to maintain backwards compatibility with older log messages.  In the new implementation, we're going to replace info-level logging with tracking log events.  This is a good opportunity to simplify the code.

For now, please remove `_log_embargo_reason` and the `partial` code.  I'd recommend breaking this into a small number of helper functions called by this function, each of which handle one particular check.  For example:
```
     # Check the global IP whitelist/blacklist
     if _check_ip_lists(ip_address):
          # ...
     
     user_country_from_ip = ...
     if _check_user_country(user_country_from_ip, course_id):
          #...

     user_country_from_profile = ...
     if _check_user_country(user_country_from_profile, course_id):
          # ...
```

I think this structure will make the code easier to follow.  It will also make it easy to add the appropriate tracking events in a future story, since each case will be a separate "if" statement.